craft_parts/executor/part_handler.py

```diff
@@ -0,0 +1,387 @@
+        #
+        #   1. So cleaning and re-running the stage step works even if `organize` is
+        #      used
+        #   2. So collision detection takes organization into account, i.e. we can use
+        #      organization to get around file collisions between parts when staging.
+        #
+        # If `update` is true, we give the snapcraft CLI permission to overwrite files
+        # that already exist. Typically we do NOT want this, so that parts don't
+        # accidentally clobber e.g. files brought in from stage-packages, but in the
+        # case of updating build, we want the part to have the ability to organize over
+        # the files it organized last time around. We can be confident that this won't
+        # overwrite anything else, because to do so would require changing the
+        # `organize` keyword, which will make the build step dirty and require a clean
+        # instead of an update.
+        self._organize(overwrite=update)
+
+        assets = {
+            "build-packages": self._build_packages,
+            "build-snaps": self._build_snaps,
+        }
+        assets.update(_get_machine_manifest())
+
+        state = states.BuildState(
+            part_properties=self._part_properties,
+            project_options=step_info.project_options,
+            assets=assets,
+        )
+        return state
+
+    def _run_stage(self, step_info: StepInfo) -> StepState:
+        self._make_dirs()
+
+        contents = self._run_step(
+            step_info=step_info,
+            scriptlet_name="override-stage",
+            work_dir=self._part.stage_dir,
+        )
+
+        state = states.StageState(
+            part_properties=self._part_properties,
+            project_options=step_info.project_options,
+            files=contents.files,
+            directories=contents.dirs,
+        )
+        return state
+
+    def _run_prime(self, step_info: StepInfo) -> StepState:
+        self._make_dirs()
+
+        contents = self._run_step(
+            step_info=step_info,
+            scriptlet_name="override-prime",
+            work_dir=self._part.prime_dir,
+        )
+
+        state = states.PrimeState(
+            part_properties=self._part_properties,
+            project_options=step_info.project_options,
+            files=contents.files,
+            directories=contents.dirs,
+        )
+        return state
+
+    def _run_step(
+        self, *, step_info: StepInfo, scriptlet_name: str, work_dir: Path
+    ) -> StepContents:
+        """Run the scriptlet if overriding, otherwise run the built-in handler.
+
+        :param step_info: Information about the step to execute.
+        :param scriptlet_name: The name of this step's scriptlet.
+        :param work_dir: The path to run the scriptlet on.
+
+        :return: If step is Stage or Prime, return a tuple of sets containing
+            the step's file and directory artifacts.
+        """
+        if not step_info.step:
```

Comment:
the type definition of `StepInfo` does not allow for None step, seems an unnecessary check unless it is set to None somewhere (and shouldn't be), though harmless

Reply:
This is just a consistency check to make sure nobody called us with a (handcrafted?) invalid parameter, and it also makes pyright happy.

Reply:
Interesting, that's what I was wondering.  Why does pyright think it can be None?

Reply:
I just used the current set of linting tools after removing the check and nobody complained, so it seems that we can remove it now.